nemo/collections/nlp/modules/common/text_generation_strategy.py

```diff
@@ -147,26 +155,22 @@ def end_of_generation_condition(
         returns:
             a boolean tensor indicating whether the generation should stop
         """
-        if len(end_strings) == 1 and end_strings[0] == END_OF_SEQ:
+        if (len(end_strings) == 1 and end_strings[0] == END_OF_SEQ) or not end_strings:
+            # Simple scenario: only finish on end of document token.
             return prev == eod_id
-        else:
-            tokenizer = self.model.tokenizer
-            conditions = []
-            end_tokens = set()
-            end_tokens.add(eod_id)
-            for end_string in end_strings:
-                ids_1 = tokenizer.text_to_ids(f'<extra_id_1>{end_string}')
-                ids_2 = tokenizer.text_to_ids('<extra_id_1>')
-                if len(ids_1) <= len(ids_2):
-                    continue
-                token_id = ids_1[len(ids_2) :][0]
```

Comment:
Just to be sure we're on the same page, I think you mean "why don't we check that `len(ids_1) == len(ids_2) + 1` (because `token_id` in that code is an integer so it has no length)

This indeed fixes the bug, but there is still a need to optimize this function because calling `text = tokenizer.ids_to_text(token_item.tolist())` below is quite inefficient (we are decoding the whole model output at each step, over the whole batch).

The optimization part of this PR has two parts:

1. (Most important) Try to skip the text decoding if possible, by only comparing `prev` to the list of end tokens
2. (Less important) Cache the list of end tokens / end strings so that we don't re-compute it at each step. This matters less and I implemented it mostly to be safe in all cases, especially when using a tokenizer that would somehow be pretty slow (compared to the model generation speed). I can get rid of it if you feel like this isn't worth it (NB: I tried to keep the complexity within `_get_end_of_generation_tokens_and_strings(()` and simplify `end_of_generation_condition()`)

Reply:
NB: also we can't cache `self.end_tokens` / `self.end_strings` because the API of the method supports arbitrary values for `end_strings`, so it would be dangerous to cache these values (unless we use `end_strings` for the cache key -- I thought about it but did not see an obvious advantage so I didn't go for this option).

Reply:
I'm fine with whichever way we cache, so all good there

okay, I see so (1) brings the major speed improvement here, makes sense. Added a new comment below that can help further improve speed

Reply:
Alright, resolving this comment then